File: datalayer_core/sdk/auth/strategies.py
# Copyright (c) 2023-2025 Datalayer, Inc.
# Distributed under the terms of the Modified BSD License.

"""
Copyright (c) 2023-2025 Datalayer, Inc.

Datalayer License
"""

"""Authentication strategy implementations for Python SDK."""

from abc import ABC, abstractmethod
from typing import Any, Dict, Optional, Tuple

import httpx

from .storage import TokenStorage


class AuthStrategy(ABC):
    """Base authentication strategy interface."""

    def __init__(self, iam_url: str, storage: Optional[TokenStorage] = None):
        """Initialize strategy.

        Args:
            iam_url: IAM service URL
            storage: Optional token storage backend
        """
        self.iam_url = iam_url
        self.storage = storage

    @abstractmethod
    async def authenticate(self, **kwargs: Any) -> Tuple[Dict[str, Any], str]:
        """
        Authenticate and return (user, token) tuple.

        Returns
        -------
        tuple of (dict, str)
            Tuple of (user_dict, token_string).

        Raises
        ------
        Exception
            If authentication fails
        """
        pass

    @abstractmethod
    def can_handle(self, **kwargs: Any) -> bool:
        """
        Check if this strategy can handle the given options.

        Returns
        -------
        bool
            True if this strategy can authenticate with given options.
        """
        pass

    async def _validate_token(self, token: str) -> Dict[str, Any]:
        """
        Validate a token by calling whoami API.

        Parameters
        ----------
        token : str
            Token to validate.

        Returns
        -------
        dict
            User dictionary.

        Raises
        ------
        Exception
            If token is invalid
        """
        async with httpx.AsyncClient() as client:
            response = await client.get(
                f"{self.iam_url}/api/iam/v1/whoami",
                headers={"Authorization": f"Bearer {token}"},
            )
            response.raise_for_status()
            data = response.json()

            # The whoami endpoint returns {success, message, profile}
            if not data or "profile" not in data:
                raise ValueError("Invalid response from whoami API")

            return data["profile"]


class TokenAuthStrategy(AuthStrategy):
    """Token-based authentication strategy."""

    def can_handle(self, **kwargs: Any) -> bool:
        """Check if token is provided."""
        return "token" in kwargs and kwargs["token"] is not None

    async def authenticate(self, **kwargs: Any) -> Tuple[Dict[str, Any], str]:
        """Authenticate using existing token."""
        token = kwargs.get("token")
        if not token:
            raise ValueError("Token is required for token authentication")

        # Validate the token
        user = await self._validate_token(token)

        # Store token if requested
        no_store = kwargs.get("no_store", False)
        if not no_store and self.storage:
            self.storage.set_token(token)

        return user, token


class CredentialsAuthStrategy(AuthStrategy):
    """Credentials-based authentication strategy."""

    def can_handle(self, **kwargs: Any) -> bool:
        """Check if handle and password are provided."""
        return (
            "handle" in kwargs
            and kwargs["handle"]
            and "password" in kwargs
            and kwargs["password"]
        )

    async def authenticate(self, **kwargs: Any) -> Tuple[Dict[str, Any], str]:
        """Authenticate using handle and password."""
        handle = kwargs.get("handle")
        password = kwargs.get("password")

        if not handle or not password:
            raise ValueError(
                "Handle and password are required for credentials authentication"
            )

        # Call login API
        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{self.iam_url}/api/iam/v1/authentication/login",
                json={"handle": handle, "password": password},
            )
            response.raise_for_status()
            data = response.json()

            if not data or not data.get("success") or not data.get("token"):
                raise ValueError(data.get("message", "Login failed"))

            token = data["token"]

        # Get user profile
        user = await self._validate_token(token)

        # Store token if requested
        no_store = kwargs.get("no_store", False)
        if not no_store and self.storage:
            self.storage.set_token(token)

        return user, token


class StorageAuthStrategy(AuthStrategy):
    """Storage-based authentication strategy."""

    def can_handle(self, **kwargs: Any) -> bool:
        """Check if storage has a token."""
        if not self.storage or not self.storage.is_available():
            return False
        token = self.storage.get_token()
        return token is not None

    async def authenticate(self, **kwargs: Any) -> Tuple[Dict[str, Any], str]:
        """Authenticate using token from storage."""
        if not self.storage:
            raise ValueError("Storage is required for storage-based authentication")

        token = self.storage.get_token()
        if not token:
            raise ValueError("No token found in storage")

        # Validate the token
        user = await self._validate_token(token)

        return user, token


class BrowserAuthStrategy(AuthStrategy):
    """Browser OAuth authentication strategy."""

    def can_handle(self, **kwargs: Any) -> bool:
        """Check if browser auth is requested."""
        return kwargs.get("use_browser", False)

    async def authenticate(self, **kwargs: Any) -> Tuple[Dict[str, Any], str]:
        """Authenticate using browser OAuth flow."""
        # TODO: Implement browser OAuth flow
        # This would involve:
        # 1. Starting local HTTP server
        # 2. Opening browser with OAuth URL
        # 3. Handling callback with authorization code
        # 4. Exchanging code for token
        raise NotImplementedError("Browser OAuth authentication not yet implemented")

File: src/client/auth/strategies.ts
/*
 * Copyright (c) 2023-2025 Datalayer, Inc.
 * Distributed under the terms of the Modified BSD License.
 */

/*
 * Copyright (c) 2023-2025 Datalayer, Inc.
 *
 * Datalayer License
 */

/**
 * Authentication strategy implementations
 */

import * as authentication from '../../api/iam/authentication';
import * as profile from '../../api/iam/profile';
import type {
  AuthStrategy,
  AuthOptions,
  AuthResult,
  TokenStorage,
} from './types';
import { UserDTO } from '../../models/UserDTO';

/**
 * Base authentication strategy with common functionality
 */
abstract class BaseAuthStrategy implements AuthStrategy {
  constructor(
    protected iamRunUrl: string,
    protected storage?: TokenStorage,
  ) {}

  abstract authenticate(options: AuthOptions): Promise<AuthResult>;
  abstract canHandle(options: AuthOptions): boolean;

  /**
   * Validate a token by calling whoami
   */
  protected async validateToken(token: string): Promise<UserDTO> {
    const response = await profile.whoami(token, this.iamRunUrl);

    if (!response || !response.profile) {
      throw new Error('Invalid response from profile API');
    }

    const userData = {
      id: response.profile.id,
      uid: response.profile.uid,
      handle_s: response.profile.handle_s || response.profile.handle,
      email_s: response.profile.email_s || response.profile.email,
      first_name_t:
        response.profile.first_name_t || response.profile.first_name || '',
      last_name_t:
        response.profile.last_name_t || response.profile.last_name || '',
      avatar_url_s:
        response.profile.avatar_url_s || response.profile.avatar_url,
    };

    return new UserDTO(userData, undefined);
  }
}

/**
 * Token-based authentication strategy
 * Authenticates using an existing token
 */
export class TokenAuthStrategy extends BaseAuthStrategy {
  canHandle(options: AuthOptions): boolean {
    return !!options.token;
  }

  async authenticate(options: AuthOptions): Promise<AuthResult> {
    if (!options.token) {
      throw new Error('Token is required for token authentication');
    }

    // Validate the token
    const user = await this.validateToken(options.token);

    // Store the token if requested
    if (!options.noStore && this.storage) {
      if (this.storage.setToken) {
        await this.storage.setToken(options.token);
      }
      if (this.storage.setUser) {
        await this.storage.setUser(user);
      }
    }

    return { user, token: options.token };
  }
}

/**
 * Credentials-based authentication strategy
 * Authenticates using handle and password
 */
export class CredentialsAuthStrategy extends BaseAuthStrategy {
  canHandle(options: AuthOptions): boolean {
    return !!options.handle && !!options.password;
  }

  async authenticate(options: AuthOptions): Promise<AuthResult> {
    if (!options.handle || !options.password) {
      throw new Error(
        'Handle and password are required for credentials authentication',
      );
    }

    // Call the login API
    const response = await authentication.login(
      { handle: options.handle, password: options.password },
      this.iamRunUrl,
    );

    if (!response || !response.success || !response.token) {
      throw new Error(response?.message || 'Login failed');
    }

    const token = response.token;

    // Get user profile
    const user = await this.validateToken(token);

    // Store the token if requested
    if (!options.noStore && this.storage) {
      this.storage.setToken?.(token);
      this.storage.setUser?.(user);
    }

    return { user, token };
  }
}

/**
 * Storage-based authentication strategy
 * Authenticates using a token from storage
 */
export class StorageAuthStrategy extends BaseAuthStrategy {
  canHandle(options: AuthOptions): boolean {
    // Can handle if:
    // 1. Storage is available
    // 2. No other auth method is provided (this is the fallback strategy)
    if (!this.storage || !this.storage.isAvailable()) {
      return false;
    }

    // Return true if no explicit auth method provided (empty options or noStore only)
    // This allows the authenticate() method to check storage async
    const hasExplicitAuthMethod =
      !!options.token ||
      !!options.handle ||
      !!options.password ||
      !!options.useBrowser;

    return !hasExplicitAuthMethod;
  }

  async authenticate(options: AuthOptions): Promise<AuthResult> {
    if (!this.storage) {
      throw new Error('Storage is required for storage-based authentication');
    }

    // Try async getTokenAsync first (for VS Code async keytar support)
    let token: string | null = null;
    if (
      'getTokenAsync' in this.storage &&
      typeof (this.storage as any).getTokenAsync === 'function'
    ) {
      token = await (this.storage as any).getTokenAsync();
    } else if (this.storage.getToken) {
      token = this.storage.getToken();
    }

    if (!token) {
      throw new Error('No token found in storage');
    }

    // Validate the token
    const user = await this.validateToken(token);

    return { user, token };
  }
}

/**
 * Browser OAuth strategy
 * Authenticates using browser-based OAuth flow with GitHub or LinkedIn
 */
export class BrowserOAuthStrategy extends BaseAuthStrategy {
  canHandle(options: AuthOptions): boolean {
    return !!options.useBrowser;
  }

  async authenticate(options: AuthOptions): Promise<AuthResult> {
    // Import OAuth2 API
    const { getOAuth2AuthzUrl } = await import('../../api/iam/oauth2');

    // Default to GitHub provider
    const provider = options.oauthProvider || 'github';

    if (provider !== 'github' && provider !== 'linkedin') {
      throw new Error(
        `Unsupported OAuth provider: ${provider}. Use 'github' or 'linkedin'.`,
      );
    }

    // Generate callback URI
    const callbackUri =
      options.callbackUri || `${window.location.origin}/auth/callback`;

    try {
      // Step 1: Get OAuth authorization URL
      const authzResponse = await getOAuth2AuthzUrl(
        provider,
        callbackUri,
        this.iamRunUrl,
        options.nonce,
      );

      // Step 2: Open OAuth URL in popup or redirect
      if (options.usePopup !== false) {
        // Use popup window
        const popup = window.open(
          authzResponse.loginURL,
          'oauth-login',
          'width=600,height=700,left=100,top=100',
        );

        if (!popup) {
          throw new Error(
            'Failed to open OAuth popup. Please allow popups for this site.',
          );
        }

        // Wait for callback - Datalayer redirects to our callback with ?user=<json>&token=<token>
        const result = await this.waitForOAuthCallback(popup, callbackUri);

        // Extract token and user from callback result
        const token = result.token;
        const user = result.user;

        if (!token) {
          throw new Error('No token received from OAuth callback');
        }

        if (!user) {
          throw new Error('No user data received from OAuth callback');
        }

        // Store token if requested
        if (!options.noStore && this.storage) {
          this.storage.setToken?.(token);
          this.storage.setUser?.(user);
        }

        return { user, token };
      } else {
        // Redirect to OAuth URL
        window.location.href = authzResponse.loginURL;

        // Return a pending promise (page will redirect)
        return new Promise(() => {
          // This promise never resolves because we redirect
        });
      }
    } catch (error) {
      throw new Error(
        `OAuth authentication failed: ${error instanceof Error ? error.message : String(error)}`,
      );
    }
  }

  /**
   * Wait for OAuth callback in popup window
   * Expects: { user, token, error } from the Datalayer OAuth callback
   */
  private async waitForOAuthCallback(
    popup: Window,
    callbackUri: string,
  ): Promise<{ user?: any; token?: string; error?: string }> {
    return new Promise((resolve, reject) => {
      const timeout = setTimeout(
        () => {
          popup.close();
          reject(new Error('OAuth authentication timed out'));
        },
        5 * 60 * 1000,
      ); // 5 minute timeout

      // Listen for message from popup
      const handleMessage = (event: MessageEvent) => {
        // Verify origin
        if (event.origin !== new URL(callbackUri).origin) {
          return;
        }

        clearTimeout(timeout);
        window.removeEventListener('message', handleMessage);
        popup.close();

        if (event.data.error) {
          reject(new Error(`OAuth error: ${event.data.error}`));
        } else {
          resolve(event.data);
        }
      };

      window.addEventListener('message', handleMessage);

      // Check if popup was closed manually
      const checkClosed = setInterval(() => {
        if (popup.closed) {
          clearInterval(checkClosed);
          clearTimeout(timeout);
          window.removeEventListener('message', handleMessage);
          reject(new Error('OAuth popup was closed'));
        }
      }, 1000);
    });
  }
}
